[C10D] support group_src/dst in broadcast/reduce ops

wconstab opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #140847
* __->__ #140843

Also add mypy annotations

Partially addresses RFC 0042 (pytorch/rfcs#71)
See more details/motivation in #140460

cc @H-Huang @awgu @kwen2501 @wanchaol @fegin @fduwjj @wz337 @d4l3k @c-p-i-o

wconstab committed:
[C10D] support group_src/dst in broadcast/reduce ops

Also add mypy annotations

[ghstack-poisoned]

wconstab committed:
Update on "[C10D] support group_src/dst in broadcast/reduce ops"


Also add mypy annotations

cc H-Huang awgu kwen2501 wanchaol fegin fduwjj wz337 d4l3k c-p-i-o

[ghstack-poisoned]

wconstab committed:
Update on "[C10D] support group_src/dst in broadcast/reduce ops"


Also add mypy annotations

cc H-Huang awgu kwen2501 wanchaol fegin fduwjj wz337 d4l3k c-p-i-o

[ghstack-poisoned]

wconstab committed:
Update on "[C10D] support group_src/dst in broadcast/reduce ops"


Also add mypy annotations

cc H-Huang awgu kwen2501 wanchaol fegin fduwjj wz337 d4l3k c-p-i-o

[ghstack-poisoned]

wconstab committed:
Update on "[C10D] support group_src/dst in broadcast/reduce ops"


Also add mypy annotations

cc H-Huang awgu kwen2501 wanchaol fegin fduwjj wz337 d4l3k c-p-i-o

[ghstack-poisoned]

wconstab committed:
Update on "[C10D] support group_src/dst in broadcast/reduce ops"


Also add mypy annotations

[ghstack-poisoned]

wconstab committed:
Update on "[C10D] support group_src/dst in broadcast/reduce ops"


Also add mypy annotations

[ghstack-poisoned]

wconstab committed:
Update on "[C10D] support group_src/dst in broadcast/reduce ops"


Also add mypy annotations

Partially addresses RFC 0042 (pytorch/rfcs#71)
See more details/motivation in #140460

[ghstack-poisoned]

wconstab committed:
Update on "[C10D] support group_src/dst in broadcast/reduce ops"


Also add mypy annotations

Partially addresses RFC 0042 (pytorch/rfcs#71)
See more details/motivation in #140460

[ghstack-poisoned]

kwen2501 reviewed (approved):
Code looks a lot cleaner now! Great work!

wconstab commented on test/distributed/test_c10d_nccl.py:
```diff
@@ -3893,20 +3900,27 @@ def test_send_recv_subgroup(self, async_op, group_rank):
+            if group_rank:
+                c10d.broadcast(x, group_src=1, group=subgroup)
+            else:
+                c10d.broadcast(x, src=self.rank + 1, group=subgroup)
```
is it? I saw a 'world_size=4' hardcode and a `if self.rank >= world_size: return` so i thought it was correct even though it insists on using exactly 4 gpus rather than being flexible to all available gpus

wconstab commented on torch/distributed/distributed_c10d.py:
```diff
@@ -3331,6 +3345,11 @@ def broadcast_object_list(object_list, src=0, group=None, device=None):
+    group = _group_or_default_group(group)
+    if src is None and group_src is None:
+        src = 0
+    global_src = _canonicalize_group_rank(group, src, group_src, return_global=True)
+    group_src = _canonicalize_group_rank(group, src, group_src, return_global=False)
```
wait, you're right i think i can delete group_src here.  Seems totally unused.  Maybe i refactored more later and forgot to clean up.

wconstab committed:
Update on "[C10D] support group_src/dst in broadcast/reduce ops"


Also add mypy annotations

Partially addresses RFC 0042 (pytorch/rfcs#71)
See more details/motivation in #140460

[ghstack-poisoned]

kwen2501 reviewed (approved):
Change looks good.
